File: java/design_pattern/mediator/MainMediator.java
/**
 * Comprehensive demonstration of the Mediator Pattern in Java
 *
 * The Mediator pattern defines an object that encapsulates how a set of objects
 * interact. It promotes loose coupling by keeping objects from referring to
 * each other explicitly.
 */

import java.util.ArrayList;
import java.util.List;

// Mediator interface
interface ChatMediator {
    void sendMessage(String message, User user);
    void addUser(User user);
}

// Concrete Mediator
class ChatRoom implements ChatMediator {
    private List<User> users = new ArrayList<>();

    @Override
    public void addUser(User user) {
        users.add(user);
        System.out.println("  [ChatRoom] " + user.getName() + " joined the chat");
    }

    @Override
    public void sendMessage(String message, User sender) {
        for (User user : users) {
            if (user != sender) {
                user.receive(message, sender.getName());
            }
        }
    }
}

// Colleague
abstract class User {
    protected ChatMediator mediator;
    protected String name;

    public User(ChatMediator mediator, String name) {
        this.mediator = mediator;
        this.name = name;
    }

    public String getName() { return name; }

    public abstract void send(String message);
    public abstract void receive(String message, String from);
}

// Concrete Colleagues
class ChatUser extends User {
    public ChatUser(ChatMediator mediator, String name) {
        super(mediator, name);
    }

    @Override
    public void send(String message) {
        System.out.println("  [" + name + "] Sending: " + message);
        mediator.sendMessage(message, this);
    }

    @Override
    public void receive(String message, String from) {
        System.out.println("  [" + name + "] Received from " + from + ": " + message);
    }
}

// Another example: Air Traffic Control

interface ATCMediator {
    void registerFlight(Flight flight);
    boolean requestLanding(Flight flight);
    void notifyDeparture(Flight flight);
}

class AirTrafficControl implements ATCMediator {
    private List<Flight> flights = new ArrayList<>();
    private boolean runwayFree = true;

    @Override
    public void registerFlight(Flight flight) {
        flights.add(flight);
        System.out.println("  [ATC] Registered flight: " + flight.getFlightNumber());
    }

    @Override
    public boolean requestLanding(Flight flight) {
        if (runwayFree) {
            runwayFree = false;
            System.out.println("  [ATC] " + flight.getFlightNumber() + " cleared for landing");
            return true;
        } else {
            System.out.println("  [ATC] " + flight.getFlightNumber() + " please hold, runway busy");
            return false;
        }
    }

    @Override
    public void notifyDeparture(Flight flight) {
        runwayFree = true;
        System.out.println("  [ATC] " + flight.getFlightNumber() + " has departed, runway is free");
        // Notify other waiting flights
        for (Flight f : flights) {
            if (f != flight && f.isWaiting()) {
                f.notifyRunwayFree();
            }
        }
    }
}

class Flight {
    private ATCMediator atc;
    private String flightNumber;
    private boolean waiting = false;

    public Flight(ATCMediator atc, String flightNumber) {
        this.atc = atc;
        this.flightNumber = flightNumber;
        atc.registerFlight(this);
    }

    public String getFlightNumber() { return flightNumber; }
    public boolean isWaiting() { return waiting; }

    public void requestLanding() {
        System.out.println("  [" + flightNumber + "] Requesting landing");
        if (!atc.requestLanding(this)) {
            waiting = true;
        }
    }

    public void notifyRunwayFree() {
        System.out.println("  [" + flightNumber + "] Notified runway is free, requesting landing");
        waiting = false;
        atc.requestLanding(this);
    }

    public void depart() {
        System.out.println("  [" + flightNumber + "] Taking off");
        atc.notifyDeparture(this);
    }
}

// Third example: Smart Home Mediator

interface SmartHomeMediator {
    void notify(SmartDevice sender, String event);
}

abstract class SmartDevice {
    protected SmartHomeMediator mediator;
    protected String name;

    public SmartDevice(SmartHomeMediator mediator, String name) {
        this.mediator = mediator;
        this.name = name;
    }

    public String getName() { return name; }
}

class SmartLight extends SmartDevice {
    private boolean on = false;

    public SmartLight(SmartHomeMediator mediator, String name) {
        super(mediator, name);
    }

    public void turnOn() {
        on = true;
        System.out.println("  [" + name + "] Light is ON");
        mediator.notify(this, "LIGHT_ON");
    }

    public void turnOff() {
        on = false;
        System.out.println("  [" + name + "] Light is OFF");
        mediator.notify(this, "LIGHT_OFF");
    }
}

class Thermostat extends SmartDevice {
    private int temperature = 70;

    public Thermostat(SmartHomeMediator mediator, String name) {
        super(mediator, name);
    }

    public void setTemperature(int temp) {
        temperature = temp;
        System.out.println("  [" + name + "] Temperature set to " + temp + "°F");
        mediator.notify(this, "TEMP_CHANGE");
    }

    public int getTemperature() { return temperature; }
}

class SecuritySystem extends SmartDevice {
    private boolean armed = false;

    public SecuritySystem(SmartHomeMediator mediator, String name) {
        super(mediator, name);
    }

    public void arm() {
        armed = true;
        System.out.println("  [" + name + "] Security system ARMED");
        mediator.notify(this, "SECURITY_ARMED");
    }

    public void disarm() {
        armed = false;
        System.out.println("  [" + name + "] Security system DISARMED");
        mediator.notify(this, "SECURITY_DISARMED");
    }
}

class CoffeeMaker extends SmartDevice {
    public CoffeeMaker(SmartHomeMediator mediator, String name) {
        super(mediator, name);
    }

    public void startBrewing() {
        System.out.println("  [" + name + "] Starting to brew coffee");
    }
}

class SmartHomeController implements SmartHomeMediator {
    private SmartLight light;
    private Thermostat thermostat;
    private SecuritySystem security;
    private CoffeeMaker coffeeMaker;

    public void setDevices(SmartLight light, Thermostat thermostat,
                          SecuritySystem security, CoffeeMaker coffeeMaker) {
        this.light = light;
        this.thermostat = thermostat;
        this.security = security;
        this.coffeeMaker = coffeeMaker;
    }

    @Override
    public void notify(SmartDevice sender, String event) {
        switch (event) {
            case "SECURITY_DISARMED":
                // When security is disarmed (someone came home)
                System.out.println("  [Controller] Welcome home! Adjusting settings...");
                light.turnOn();
                thermostat.setTemperature(72);
                coffeeMaker.startBrewing();
                break;

            case "SECURITY_ARMED":
                // When security is armed (leaving home)
                System.out.println("  [Controller] Goodbye! Saving energy...");
                light.turnOff();
                thermostat.setTemperature(65);
                break;
        }
    }
}

public class MainMediator {
    public static void main(String[] args) {
        System.out.println("=== Mediator Pattern Demonstration ===\n");

        // Chat room example
        System.out.println("--- 1. Chat Room Mediator ---");

        ChatMediator chatRoom = new ChatRoom();

        User alice = new ChatUser(chatRoom, "Alice");
        User bob = new ChatUser(chatRoom, "Bob");
        User charlie = new ChatUser(chatRoom, "Charlie");

        chatRoom.addUser(alice);
        chatRoom.addUser(bob);
        chatRoom.addUser(charlie);

        System.out.println();
        alice.send("Hello everyone!");
        System.out.println();
        bob.send("Hi Alice!");
        System.out.println();

        // Air Traffic Control example
        System.out.println("--- 2. Air Traffic Control Mediator ---");

        ATCMediator atc = new AirTrafficControl();

        Flight flight1 = new Flight(atc, "AA123");
        Flight flight2 = new Flight(atc, "UA456");
        Flight flight3 = new Flight(atc, "DL789");

        System.out.println();
        flight1.requestLanding();
        flight2.requestLanding();  // Will be put on hold
        flight3.requestLanding();  // Will be put on hold

        System.out.println();
        flight1.depart();  // Frees runway, notifies waiting flights
        System.out.println();

        // Smart Home example
        System.out.println("--- 3. Smart Home Mediator ---");

        SmartHomeController controller = new SmartHomeController();

        SmartLight light = new SmartLight(controller, "Living Room Light");
        Thermostat thermostat = new Thermostat(controller, "Main Thermostat");
        SecuritySystem security = new SecuritySystem(controller, "Home Security");
        CoffeeMaker coffeeMaker = new CoffeeMaker(controller, "Kitchen Coffee Maker");

        controller.setDevices(light, thermostat, security, coffeeMaker);

        System.out.println("\nComing home (disarming security):");
        security.disarm();

        System.out.println("\nLeaving home (arming security):");
        security.arm();

        System.out.println("\n=== Summary ===");
        System.out.println("Mediator pattern benefits:");
        System.out.println("  - Reduces coupling between colleagues");
        System.out.println("  - Centralizes complex communication logic");
        System.out.println("  - Simplifies object protocols");
        System.out.println("  - Easier to understand object interactions");
        System.out.println("\nConsiderations:");
        System.out.println("  - Mediator can become complex (God Object)");
        System.out.println("  - One-to-many becomes many-to-one");
    }
}

File: python/design_pattern/mediator/main_mediator.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
Mediator Pattern

The Mediator pattern defines an object that encapsulates how a set of objects
interact. It promotes loose coupling by keeping objects from referring to each
other explicitly and lets you vary their interaction independently.

Key Components:
- Mediator: Interface for communicating with Colleague objects
- ConcreteMediator: Implements cooperation between Colleagues
- Colleague: Communicates with other Colleagues via Mediator
"""

from abc import ABC, abstractmethod
from datetime import datetime


# Mediator interface
class ChatMediator(ABC):
    """Abstract mediator for chat room."""

    @abstractmethod
    def send_message(self, message: str, sender: "User") -> None:
        pass

    @abstractmethod
    def send_private_message(
        self, message: str, sender: "User", recipient: "User"
    ) -> None:
        pass

    @abstractmethod
    def add_user(self, user: "User") -> None:
        pass

    @abstractmethod
    def remove_user(self, user: "User") -> None:
        pass


# Colleague base class
class User(ABC):
    """Abstract colleague representing a chat user."""

    def __init__(self, name: str, mediator: ChatMediator):
        self._name = name
        self._mediator = mediator
        self._messages: list[str] = []

    @property
    def name(self) -> str:
        return self._name

    def send(self, message: str) -> None:
        """Send message to all users via mediator."""
        timestamp = datetime.now().strftime("%H:%M:%S")
        print(f"[{timestamp}] {self._name} sends: {message}")
        self._mediator.send_message(message, self)

    def send_private(self, message: str, recipient: "User") -> None:
        """Send private message via mediator."""
        timestamp = datetime.now().strftime("%H:%M:%S")
        print(f"[{timestamp}] {self._name} -> {recipient.name} (private): {message}")
        self._mediator.send_private_message(message, self, recipient)

    @abstractmethod
    def receive(self, message: str, sender: "User") -> None:
        """Receive message from another user."""
        pass

    def get_messages(self) -> list[str]:
        return self._messages.copy()


# Concrete Colleagues
class RegularUser(User):
    """Regular chat user."""

    def receive(self, message: str, sender: "User") -> None:
        formatted = f"{sender.name}: {message}"
        self._messages.append(formatted)
        print(f"  [{self._name} received] {formatted}")


class PremiumUser(User):
    """Premium user with special formatting."""

    def receive(self, message: str, sender: "User") -> None:
        formatted = f"★ {sender.name}: {message}"
        self._messages.append(formatted)
        print(f"  [{self._name} received] {formatted}")


class BotUser(User):
    """Bot user that can auto-respond."""

    def __init__(self, name: str, mediator: ChatMediator):
        super().__init__(name, mediator)
        self._auto_responses: dict[str, str] = {
            "help": "Available commands: !help, !time, !users",
            "time": f"Current time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}",
        }

    def receive(self, message: str, sender: "User") -> None:
        formatted = f"{sender.name}: {message}"
        self._messages.append(formatted)
        print(f"  [{self._name} received] {formatted}")

        # Auto-respond to commands
        if message.startswith("!"):
            cmd = message[1:].lower()
            if cmd in self._auto_responses:
                self.send(self._auto_responses[cmd])


# Concrete Mediator
class ChatRoom(ChatMediator):
    """Concrete mediator implementing chat room."""

    def __init__(self, name: str):
        self._name = name
        self._users: list[User] = []
        self._message_history: list[dict] = []

    def add_user(self, user: User) -> None:
        self._users.append(user)
        print(f">>> {user.name} joined {self._name}")
        # Notify other users
        for u in self._users:
            if u != user:
                u.receive(f"{user.name} has joined the chat", BotUser("System", self))

    def remove_user(self, user: User) -> None:
        if user in self._users:
            self._users.remove(user)
            print(f"<<< {user.name} left {self._name}")
            # Notify other users
            for u in self._users:
                u.receive(f"{user.name} has left the chat", BotUser("System", self))

    def send_message(self, message: str, sender: User) -> None:
        """Broadcast message to all users except sender."""
        self._message_history.append(
            {
                "timestamp": datetime.now(),
                "sender": sender.name,
                "message": message,
                "type": "broadcast",
            }
        )

        for user in self._users:
            if user != sender:
                user.receive(message, sender)

    def send_private_message(self, message: str, sender: User, recipient: User) -> None:
        """Send private message to specific user."""
        self._message_history.append(
            {
                "timestamp": datetime.now(),
                "sender": sender.name,
                "recipient": recipient.name,
                "message": message,
                "type": "private",
            }
        )

        if recipient in self._users:
            recipient.receive(f"(private) {message}", sender)
        else:
            sender.receive(f"User {recipient.name} is not in the chat", BotUser("System", self))

    def get_online_users(self) -> list[str]:
        return [user.name for user in self._users]

    def get_message_count(self) -> int:
        return len(self._message_history)


# Advanced Mediator Example: Air Traffic Control
class Aircraft:
    """Colleague: An aircraft communicating via ATC."""

    def __init__(self, call_sign: str, atc: "AirTrafficControl"):
        self.call_sign = call_sign
        self._atc = atc
        self._altitude = 0
        self._runway: str | None = None

    def request_landing(self) -> None:
        print(f"{self.call_sign}: Requesting permission to land")
        self._atc.request_landing(self)

    def request_takeoff(self) -> None:
        print(f"{self.call_sign}: Requesting permission for takeoff")
        self._atc.request_takeoff(self)

    def receive_clearance(self, message: str) -> None:
        print(f"{self.call_sign}: Received - {message}")

    def land(self, runway: str) -> None:
        self._runway = runway
        self._altitude = 0
        print(f"{self.call_sign}: Landing on runway {runway}")

    def takeoff(self, runway: str) -> None:
        self._runway = None
        self._altitude = 1000
        print(f"{self.call_sign}: Taking off from runway {runway}")


class AirTrafficControl:
    """Mediator: Coordinates aircraft movements."""

    def __init__(self):
        self._runways: dict[str, Aircraft | None] = {
            "27L": None,
            "27R": None,
            "09L": None,
            "09R": None,
        }
        self._landing_queue: list[Aircraft] = []
        self._takeoff_queue: list[Aircraft] = []

    def request_landing(self, aircraft: Aircraft) -> None:
        # Find available runway
        for runway, occupant in self._runways.items():
            if occupant is None:
                self._runways[runway] = aircraft
                aircraft.receive_clearance(f"Cleared to land on runway {runway}")
                aircraft.land(runway)
                return

        # No runway available, add to queue
        self._landing_queue.append(aircraft)
        position = len(self._landing_queue)
        aircraft.receive_clearance(f"Hold position, #{position} in landing queue")

    def request_takeoff(self, aircraft: Aircraft) -> None:
        # Find runway where aircraft is located
        for runway, occupant in self._runways.items():
            if occupant == aircraft:
                aircraft.receive_clearance(f"Cleared for takeoff on runway {runway}")
                aircraft.takeoff(runway)
                self._runways[runway] = None
                self._process_queue()
                return

        # Aircraft not on runway
        self._takeoff_queue.append(aircraft)
        aircraft.receive_clearance("Taxi to runway and hold")

    def _process_queue(self) -> None:
        """Process waiting aircraft when runway becomes available."""
        if self._landing_queue:
            next_aircraft = self._landing_queue.pop(0)
            self.request_landing(next_aircraft)

    def get_status(self) -> dict[str, str | None]:
        return {
            runway: (aircraft.call_sign if aircraft else "Empty")
            for runway, aircraft in self._runways.items()
        }


def main() -> None:
    print("=" * 60)
    print("Mediator Pattern - Chat Room Demo")
    print("=" * 60)

    # Demo 1: Basic chat room
    print("\n--- Chat Room Demo ---")
    chat_room = ChatRoom("General")

    # Create users
    alice = RegularUser("Alice", chat_room)
    bob = RegularUser("Bob", chat_room)
    charlie = PremiumUser("Charlie", chat_room)
    helper_bot = BotUser("HelperBot", chat_room)

    # Users join
    chat_room.add_user(alice)
    chat_room.add_user(bob)
    chat_room.add_user(charlie)
    chat_room.add_user(helper_bot)

    print("\n--- Messages ---")
    # Send messages
    alice.send("Hello everyone!")
    bob.send("Hi Alice!")
    charlie.send("How's it going?")

    # Private message
    print("\n--- Private Message ---")
    alice.send_private("Can we talk later?", bob)

    # Bot interaction
    print("\n--- Bot Interaction ---")
    bob.send("!help")

    # User leaves
    print("\n--- User Leaving ---")
    chat_room.remove_user(charlie)

    print(f"\nOnline users: {chat_room.get_online_users()}")
    print(f"Total messages: {chat_room.get_message_count()}")

    # Demo 2: Air Traffic Control
    print("\n" + "=" * 60)
    print("Mediator Pattern - Air Traffic Control Demo")
    print("=" * 60)

    atc = AirTrafficControl()

    # Create aircraft
    flight1 = Aircraft("UA123", atc)
    flight2 = Aircraft("AA456", atc)
    flight3 = Aircraft("DL789", atc)
    flight4 = Aircraft("SW101", atc)
    flight5 = Aircraft("JB202", atc)

    print("\n--- Landing Requests ---")
    flight1.request_landing()
    flight2.request_landing()
    flight3.request_landing()
    flight4.request_landing()
    flight5.request_landing()  # This one should queue

    print(f"\nRunway status: {atc.get_status()}")

    print("\n--- Takeoff Request ---")
    flight1.request_takeoff()  # This frees up a runway

    print(f"\nRunway status: {atc.get_status()}")

    print("\n" + "=" * 60)
    print("Benefits of Mediator Pattern:")
    print("=" * 60)
    print("1. Decouples colleagues from each other")
    print("2. Centralizes control logic")
    print("3. Simplifies object protocols")
    print("4. Easy to change interaction independently")
    print("5. Reduces subclassing")


if __name__ == "__main__":
    main()
